### PR TITLE
Support arbitrary HTTP clients

### PR DESCRIPTION
**Pull request by rbtcollins**
Included: surf and reqwest.

Users can implement unleash_api_client::http::HttpClient to bring
support for their own preferred client.

This permits only pulling in the relevant async ecosystem for the client
being used, which may aide compile times for users of this crate.

### CODE REVIEW

**Review by rbtcollins — COMMENTED** [review]
This is a whole-cloth rework of #31, sharing only the feature names and the pre-req refactoring commit. As such it needs a full review.

File: src/http/shim.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    type Error: Debug + Display + Error + Send + Sync + 'static;
+    type RequestBuilder;
+
+    /// Construct a HTTP client layer headername
+    fn build_header(name: &'static str) -> Result<Self::HeaderName, Self::Error>;
+
+    /// Make a get request
+    fn get(&self, uri: &str) -> Self::RequestBuilder;
+
+    /// Make a post  request
```
Comment (sighphyre) [inline src/http/shim.rs:27]:
```suggestion
    /// Make a post request
```

**Review by sighphyre — APPROVED** [review]
This looks great to my eyes! Currently it looks like while you could use another HTTP library, first class support is for Surf and Reqwest. It would be very cool if we could have an example with another HTTP client but that's low priority and should probably be done in another piece of work when we overhaul all the Unleash SDK examples

File: src/http/shim.rs (hunk 1 above)
Comment (sighphyre) [reply src/http/shim.rs:27]: Ah, didn't notice the auto merge, so this is now a moot point and really just a whitespace character. Marking this as resolved

**Pull request merged**